### PR TITLE
Implement name resolution

### PR DESCRIPTION
**Pull request by nitnelave**
Implement a very basic name resolution algorithm.

Identifiers can now be resolved or not, same for types.

The initial name resolution algorithm just checks that every variable references a variable declared before. This is obviously incomplete, as it does not include scoping or functions (without even mentioning modules and imports), and does not respect the objective that functions (and classes) should be able to reference each other without forward declaration.

---
This change is [Reviewable](https://reviewable.io/reviews/nitnelave/hopper/39)

### CODE REVIEW

File: src/ast/builtin_type.cc [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+Option<IntWidth> int_to_width(const TypeDeclaration* type) {
+  if (type == &int64) return IntWidth::W_64;
+  if (type == &int32) return IntWidth::W_32;
+  if (type == &int16) return IntWidth::W_16;
+  if (type == &int8) return IntWidth::W_8;
+  return none;
+}
+
+const TypeDeclaration* width_to_int(IntWidth width) {
```
Comment (bloody76) [inline src/ast/builtin_type.cc]: I think the naming is not very explicit here, width_to_int should return an int, but its a pointer to a TypeDeclaration, shouldn't you call it width_to_typeDec or something ?

File: src/ast/builtin_type.h [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+extern const BuiltinType void_type;
+extern const BuiltinType boolean;
+extern const BuiltinType int8;
+extern const BuiltinType int16;
+extern const BuiltinType int32;
+extern const BuiltinType int64;
+
+const TypeDeclaration* width_to_int(IntWidth width);
+Option<IntWidth> int_to_width(const TypeDeclaration* type);
```
Comment (bloody76) [inline src/ast/builtin_type.h]: same here, it seems misleading, we are not really giving a int, maybe we should name it differently.

File: test/resources/resources.cc [hunk 3]
```diff
@@ -207,13 +240,14 @@ Variant<AssertionResult, std::string> get_transformed_ir(
...
-  std::string ref_filename =
-      filename.substr(0, filename.size() - sizeof(".gh") + 1) + ".ref";
   auto first_pass = get_pretty_printed_file(filename);
   if (first_pass.is<AssertionResult>())
     return first_pass.get_unchecked<AssertionResult>();
   const std::string& first_pass_result =
       first_pass.get_unchecked<std::string>();
+
+  std::string ref_filename =
+      filename.substr(0, filename.size() - sizeof(".gh") + 1) + ".ref";
```
Comment (bloody76) [inline test/resources/resources.cc]: why not directly filename + ".ref" ? By trying to remove the .gh you force people to write .gh at the end of their filenames

Comment (bloody76) [reply test/resources/resources.cc]: Also if you want to keep that kind of operation, make a function that would be reusable

**Review by bloody76 — COMMENTED** [review]
Generally, why do you keep including cc files ? There is a recursive macro for that.

Otherwise I guess most of the comments are to discuss as I am new to the project.

**Comment by nitnelave** [comment]
Review status: 0 of 36 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[src/ast/builtin_type.cc, line 26 at r12](https://reviewable.io:443/reviews/nitnelave/hopper/39#-Kq8Y9yA8ZqNbaDPXi5q-r12-26:-Kq98QXYdKCeyXOtgNyg:bn8e66n) ([raw file](https://github.com/nitnelave/hopper/blob/2eadc8672e8d6ea88af5dcb221b8ce4f9f645994/src/ast/builtin_type.cc#L26)):*
<details><summary><i>Previously, bloody76 (Bd76) wrote…</i></summary><blockquote>

I think the naming is not very explicit here, width_to_int should return an int, but its a pointer to a TypeDeclaration, shouldn't you call it width_to_typeDec or something ?
</blockquote></details>

width_to_int_type?

---

*[src/ast/builtin_type.h, line 38 at r12](https://reviewable.io:443/reviews/nitnelave/hopper/39#-Kq8Y9yA8ZqNbaDPXi5r-r12-38:-Kq98UlNilRTzJoPxzcn:bao6pxx) ([raw file](https://github.com/nitnelave/hopper/blob/2eadc8672e8d6ea88af5dcb221b8ce4f9f645994/src/ast/builtin_type.h#L38)):*
<details><summary><i>Previously, bloody76 (Bd76) wrote…</i></summary><blockquote>

same here, it seems misleading, we are not really giving a int, maybe we should name it differently.
</blockquote></details>

This one I would argue that the input parameter is sufficiently clear. But why not int_type_to_width

---

*[test/resources/resources.cc, line 250 at r12](https://reviewable.io:443/reviews/nitnelave/hopper/39#-Kq8Y9yCyJTlSKC3ZHk0-r12-250:-Kq98k7cKtNzRt2gHmEl:b3pbm0k) ([raw file](https://github.com/nitnelave/hopper/blob/2eadc8672e8d6ea88af5dcb221b8ce4f9f645994/test/resources/resources.cc#L250)):*
<details><summary><i>Previously, bloody76 (Bd76) wrote…</i></summary><blockquote>

Also if you want to keep that kind of operation, make a function that would be reusable
</blockquote></details>

This is just in the test framework, so yes I limit to ".gh" files, because that's the format expected in the test resources. +1 for the function, though

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/nitnelave/hopper/39)*

**Comment by nitnelave** [comment]
Need to add unit tests.

---

Review status: 0 of 38 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/ast/builtin_type.cc, line 26 at r12](https://reviewable.io:443/reviews/nitnelave/hopper/39#-Kq8Y9yA8ZqNbaDPXi5q-r12-26:-KqJ9LNsPH-stJ2PjFtZ:b-896fix) ([raw file](https://github.com/nitnelave/hopper/blob/2eadc8672e8d6ea88af5dcb221b8ce4f9f645994/src/ast/builtin_type.cc#L26)):*
<details><summary><i>Previously, nitnelave (nitnelave) wrote…</i></summary><blockquote>

width_to_int_type?
</blockquote></details>

Done.

---

*[src/ast/builtin_type.h, line 38 at r12](https://reviewable.io:443/reviews/nitnelave/hopper/39#-Kq8Y9yA8ZqNbaDPXi5r-r12-38:-KqJ9MQxOfoeF9BZwons:b-896fix) ([raw file](https://github.com/nitnelave/hopper/blob/2eadc8672e8d6ea88af5dcb221b8ce4f9f645994/src/ast/builtin_type.h#L38)):*
<details><summary><i>Previously, nitnelave (nitnelave) wrote…</i></summary><blockquote>

This one I would argue that the input parameter is sufficiently clear. But why not int_type_to_width
</blockquote></details>

Done.

---

*[test/resources/resources.cc, line 250 at r12](https://reviewable.io:443/reviews/nitnelave/hopper/39#-Kq8Y9yCyJTlSKC3ZHk0-r12-250:-KqJ9osPulrmeXt_dJah:b-896fix) ([raw file](https://github.com/nitnelave/hopper/blob/2eadc8672e8d6ea88af5dcb221b8ce4f9f645994/test/resources/resources.cc#L250)):*
<details><summary><i>Previously, nitnelave (nitnelave) wrote…</i></summary><blockquote>

This is just in the test framework, so yes I limit to ".gh" files, because that's the format expected in the test resources. +1 for the function, though
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/nitnelave/hopper/39#-:-KqJBuyTDZ1w52R-asDY:b-b0fgwm)*

**Comment by bloody76** [comment]
Reviewed 11 of 21 files at r1, 1 of 1 files at r2, 1 of 1 files at r3, 3 of 4 files at r4, 2 of 6 files at r5, 7 of 7 files at r6, 4 of 5 files at r7, 1 of 1 files at r8, 3 of 4 files at r12, 3 of 5 files at r13, 12 of 12 files at r14.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/nitnelave/hopper/39)*

**Comment by bloody76** [comment]
:lgtm:

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/nitnelave/hopper/39#-:-KqT8WBDl4BR6srnIxvY:bnfp4nl)*

**Pull request merged**